04-onion-routing.md

```diff
@@ -1052,46 +1058,173 @@ The erring node then generates a new key, using the key type `ammag`.
 This key is then used to generate a pseudo-random stream, which is in turn
 applied to the packet using `XOR`.
 
-The obfuscation step is repeated by every hop along the return path.
-Upon receiving a return packet, each hop generates its `ammag`, generates the
-pseudo-random byte stream, and applies the result to the return packet before
-return-forwarding it.
-
-The origin node is able to detect that it's the intended final recipient of the
-return message, because of course, it was the originator of the corresponding
-forward packet.
-When an origin node receives an error message matching a transfer it initiated
-(i.e. it cannot return-forward the error any further) it generates the `ammag`
-and `um` keys for each hop in the route.
-It then iteratively decrypts the error message, using each hop's `ammag`
-key, and computes the HMAC, using each hop's `um` key.
-The origin node can detect the sender of the error message by matching the
-`hmac` field with the computed HMAC.
-
-The association between the forward and return packets is handled outside of
-this onion routing protocol, e.g. via association with an HTLC in a payment
-channel.
-
 Error handling for HTLCs with `path_key` is particularly fraught,
 since differences in implementations (or versions) may be leveraged to
 de-anonymize elements of the blinded path. Thus the decision turn every
 error into `invalid_onion_blinding` which will be converted to a normal
 onion error by the introduction point.
 
+### Initialization of `attribution_data`
+
+For the layout of `attribution data`, see [Removing an HTLC: `update_fulfill_htlc`, `update_fail_htlc`, and
+`update_fail_malformed_htlc`](02-peer-protocol.md#removing-an-htlc-update_fulfill_htlc-update_fail_htlc-and-update_fail_malformed_htlc).
+
+The `htlc_hold_times` field specifies the htlc hold time for each hop in units of 100 milliseconds. For example, a value
```

Comment:
I think we need to define this properly, otherwise each implementation can have its own interpretations, which can also be a privacy leak? For instance, does it start when the node receives an htlc or sends an htlc? and does it end when the node receives the failure msg or sends back the msg?

Reply:
I think with the 100 ms granularity, we are good now? I also wouldn't be sure that everyone would really implement it exactly the way we'd specify it here. The implementation complexity can vary quite a bit depending on which timing points are used.